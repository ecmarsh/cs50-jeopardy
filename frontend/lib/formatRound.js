export default function(page) {
  let title;
  switch (page) {
    case 1:
      title = 'Single';
      break;
    case 2:
      title = 'Double';
      break;
    case 3:
      title = 'Final';
      break;
    default:
      title = 'CS50';
  }

  return title + ' Jeopardy';
}
