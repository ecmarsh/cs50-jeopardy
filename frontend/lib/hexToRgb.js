export default function(hex) {
  const chunk = 2,
    radix = 16,
    rgb = [];

  let i = 1;
  while (hex && i < hex.length) {
    rgb.push(parseInt(hex.substring(i, i + chunk), radix));
    i += chunk;
  }

  return rgb.toString();
}
