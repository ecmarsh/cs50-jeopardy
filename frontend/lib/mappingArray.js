export default function(N = 0, isFilled = true, startAtZero = false) {
  if (N == 0) return [];
  if (N > 0 && !isFilled) return Array(N).fill();
  if (isFilled && !startAtZero)
    return Array.from({ length: N }, (v, k) => k + 1);
  return Array.from({ length: N }, (v, k) => k++);
}
