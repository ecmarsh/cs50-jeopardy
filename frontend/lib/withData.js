import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpoint, endpointProd } from '../config';
import { LOCAL_STATE_QUERY } from '../components/SoundEffect';
import { ANSWERED_QUESTIONS_QUERY } from '../components/GameScreen';
import gql from 'graphql-tag';

// Setup for SSR w/ next and config endpoint
function createClient({ headers }) {
  const typeDefs = gql`
    type StudyQuestion {
      data: Json
    }
  `;
  const client = new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : endpointProd,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    typeDefs,
    // cache
    clientState: {
      resolvers: {
        Mutation: {
          toggleSound(_, variables, { cache }) {
            // read the soundOn value from the cache
            const { soundOn } = cache.readQuery({
              query: LOCAL_STATE_QUERY,
            });
            // Toggle sound on || off
            const data = {
              data: { soundOn: !soundOn },
            };
            cache.writeData(data);
            return data;
          },
          addAnswered(_, { id, val }, { cache }) {
            // read the answered questions from the cache
            const { answeredQuestions } = cache.readQuery({
              query: ANSWERED_QUESTIONS_QUERY,
            });
            const newQuestion = { id, val, __typename: 'StudyQuestion' };
            const jsonified = JSON.stringify([
              newQuestion,
              ...answeredQuestions,
            ]);
            // Add new question to data
            const data = {
              data: {
                answeredQuestions: [newQuestion, ...answeredQuestions],
              },
            };

            cache.writeData(data);
            return data;
          },
          resetBoard(_, variables, { cache }) {
            // Reset answered questions
            const data = {
              data: {
                answeredQuestions: [],
              },
            };
            cache.writeData(data);
            return data;
          },
        },
      },
      defaults: {
        soundOn: false,
        answeredQuestions: [],
      },
    },
  });
  return client;
}

export default withApollo(createClient);
