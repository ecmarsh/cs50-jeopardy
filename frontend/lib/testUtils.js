import casual from 'casual';
import mappingArray from './mappingArray';

// seed it for repeatable random sequence
casual.seed(123);

const fakeQuestion = () => ({
  id: 'abc123',
  question: 'Time complexity of merge sort?',
  answer: 'n log n',
  user: {
    id: casual.uuid,
    name: 'User Name',
    __typename: 'User',
  },
  createdAt: '2019-02 - 26T19: 24: 16.000Z',
  game: {
    id: 'game123',
    name: 'GameForGeekz',
    user: {
      id: casual.uuid,
      __typename: 'User',
    },
    isPublic: casual.coin_flip,
    __typename: 'Game',
  },
  categoryQuestions: [
    {
      id: 'catQuest123',
      category: { id: 'cat123', __typename: 'Category' },
      __typename: '__CategoryQuestion',
    },
  ],
  __typename: 'Question',
});

const fakeUser = () => ({
  __typename: 'User',
  id: 'userABC123',
  name: casual.name,
  email: casual.email,
  permissions: ['USER'],
});

const fakeAdmin = () => ({
  __typename: 'User',
  id: 'adminABC123',
  name: casual.name,
  email: casual.email,
  permission: ['ADMIN', 'USER'],
  games: [],
});

const fakeCategoryQuestion = (id = 'catQuest123', difficulty = 1) => ({
  id,
  difficulty,
  isDouble: false,
  answered: casual.coin_flip,
  question: fakeQuestion(),
  category: { id: 'cat123', name: casual.word, __typename: 'Category' },
  game: { id: 'g@m3', name: 'Test Game', __typename: 'Game' },
  __typename: 'CategoryQuestion',
});

const fillCategoryQuestions = (n = 5) =>
  mappingArray(n).map(i => fakeCategoryQuestion(`catQuest-${i}`, i));

const fakeCategory = (id = 'cat123') => ({
  id,
  name: casual.word,
  game: { id: 'g@m3', name: 'Test Game' },
  categoryQuestions: fillCategoryQuestions(),
  __typename: 'Category',
});

const fillCategories = (n = 11) =>
  mappingArray(n).map(i => fakeCategory(`cat-${i}`));

const fakeTeam = id => ({
  __typename: 'Team',
  id: `t${id}`,
  name: `Team${id}`,
  score: 1000,
  game: { id: 'g@m3', name: 'Test Game' },
});

const fakeGameConfig = () => ({
  id: 'config123',
  hasDoubleJeopardy: false,
  hasRoundTimer: false,
  roundTime: 0,
  finalTime: 0,
  __typename: 'GameConfig',
});

const fakeGame = () => ({
  __typename: 'Game',
  id: 'g@m3',
  name: 'Test Game',
  createdAt: '2019-03 - 06T19: 12: 16.000Z',
  isPublic: casual.coin_flip,
  user: fakeUser(),
  teams: [fakeTeam(1), fakeTeam(2)],
  categories: fillCategories(),
  questions: [fakeQuestion()],
  config: fakeGameConfig(),
});

const fakeStudyGame = (id = 'SG123') => ({
  __typename: 'StudyGame',
  id,
  answeredQuestions: [fakeCategoryQuestion(), fakeCategoryQuestion()],
  score: 1000,
  game: fakeGame(),
  user: fakeUser(),
  createdAt: '2019-04 - 25T19: 10: 16.000Z',
});

const fakeGameSummary = () => ({
  questionsConnection: {
    aggregate: {
      count: casual.double + 1,
      __typename: 'aggregate',
    },
    __typename: 'questionsConnection',
  },
  categoryQuestionsConnection: {
    aggregate: {
      count: casual.double,
      __typename: 'aggregate',
    },
    __typename: 'categoryQuestionsConnection',
  },
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export {
  LocalStorageMock,
  fakeQuestion,
  fakeUser,
  fakeAdmin,
  fakeCategory,
  fakeCategoryQuestion,
  fakeTeam,
  fakeGame,
  fakeStudyGame,
  fakeGameSummary,
  fakeGameConfig,
};
