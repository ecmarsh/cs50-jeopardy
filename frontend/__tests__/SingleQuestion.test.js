import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import SubmittedQuestion, {
  SINGLE_QUESTION_QUERY,
} from '../components/QuestionConfirm';
import { MockedProvider } from 'react-apollo/test-utils';
import { fakeQuestion } from '../lib/testUtils';

describe('<SubmittedQuestion/>', () => {
  it('renders with submitted data', async () => {
    const mocks = [
      {
        request: {
          query: SINGLE_QUESTION_QUERY,
          variables: { id: '123' },
        },
        result: {
          data: {
            question: fakeQuestion(),
          },
        },
      },
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SubmittedQuestion id="123" />
      </MockedProvider>
    );
    // console.log(wrapper.debug());
    expect(wrapper.text()).toContain('Loading...');
    // wait for loading state to pass and update
    await wait();
    wrapper.update();
    //console.log(wrapper.debug());
    expect(toJSON(wrapper.find('h2'))).toMatchSnapshot();
    expect(toJSON(wrapper.find('span'))).toMatchSnapshot();
  });

  it('Errors with a not found question', async () => {
    const mocks = [
      {
        request: {
          query: SINGLE_QUESTION_QUERY,
          variables: { id: '123' },
        },
        result: {
          data: {
            question: null,
          },
        },
      },
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SubmittedQuestion id="123" />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const errMessage = wrapper.find('p');
    expect(wrapper.text()).toContain('Unable to load question');
    expect(toJSON(errMessage)).toMatchSnapshot();
  });
});
