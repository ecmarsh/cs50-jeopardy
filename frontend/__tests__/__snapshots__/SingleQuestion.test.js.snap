// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SubmittedQuestion/> Errors with a not found question 1`] = `
<p>
  Unable to load question.
</p>
`;

exports[`<SubmittedQuestion/> renders with submitted data 1`] = `
<h2>
  Thanks for submitting!
</h2>
`;

exports[`<SubmittedQuestion/> renders with submitted data 2`] = `
Array [
  <span
    className="Utilities__TextEmph-s11g5i-0 jQVwwM"
  >
    Time complexity of merge sort?
  </span>,
  <span
    className="Utilities__TextEmph-s11g5i-0 jQVwwM"
  >
    n log n
  </span>,
  <span
    style={
      Object {
        "padding": "1rem",
      }
    }
  >
    Â 
  </span>,
  <span
    className="Utilities__TextMuted-s11g5i-2 bdrBqE"
  >
    Submitted on:
     
    January 4, 2019 4:00 pm
  </span>,
]
`;
